File: app/src/main/java/dev/olog/basil/presentation/main/MainActivity.java
package dev.olog.basil.presentation.main;

import android.os.Bundle;
import android.view.View;

import com.github.dmstocking.optional.java.util.Optional;
import com.sothree.slidinguppanel.SlidingUpPanelLayout;

import javax.inject.Inject;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import dev.olog.basil.R;
import dev.olog.basil.presentation.base.BaseActivity;
import dev.olog.basil.presentation.navigator.Navigator;
import dev.olog.basil.presentation.widget.StoppableVerticalViewPager;
import dev.olog.basil.utils.ListUtils;

public class MainActivity extends BaseActivity {

    @Inject Navigator navigator;

    private StoppableVerticalViewPager pager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        pager = findViewById(R.id.pager);
        pager.setAdapter(new VerticalPagerAdapter(getSupportFragmentManager()));
        pager.setCurrentItem(1);
    }

    @Override
    public void onBackPressed() {
        if (pager.getCurrentItem() == 0){
            pager.setCurrentItem(1);
            return;
        } else if (pager.getCurrentItem() == 1){
            SlidingUpPanelLayout slidingPanel = findSlidingPanel();
            if (slidingPanel != null){
                if (slidingPanel.getPanelState() == SlidingUpPanelLayout.PanelState.EXPANDED){
                    slidingPanel.setPanelState(SlidingUpPanelLayout.PanelState.COLLAPSED);
                    return;
                }
            }
        }

        super.onBackPressed();
    }

    @Nullable
    private SlidingUpPanelLayout findSlidingPanel(){
        Optional<Fragment> fragment = ListUtils.find(getSupportFragmentManager().getFragments(), f -> f instanceof MainFragment);
        if (fragment.isPresent()){
            View view = fragment.get().getView();
            if (view != null){
                return view.findViewById(R.id.slidingPanel);
            }
        }
        return null;
    }

}

File: app/src/main/java/dev/olog/basil/presentation/main/MainActivity.kt
package dev.olog.basil.presentation.main

import android.os.Bundle
import com.sothree.slidinguppanel.SlidingUpPanelLayout
import dev.olog.basil.R
import dev.olog.basil.presentation.DrawsOnTop
import dev.olog.basil.presentation.base.BaseActivity
import dev.olog.basil.utils.collapse
import dev.olog.basil.utils.topMostFragment
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : BaseActivity() {

    val viewModel by lazy { MainFragmentViewModel() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        pager.adapter = VerticalPagerAdapter(supportFragmentManager)
        pager.currentItem = 1
    }

    override fun onBackPressed() {
        when {
            supportFragmentManager.topMostFragment() is DrawsOnTop -> super.onBackPressed()
            pager.currentItem == 0 -> pager.currentItem = 1
            !(pager.currentItem == 1 && findSlidingPanel()?.collapse() == true) -> {
                super.onBackPressed()
            }
            else -> super.onBackPressed()
        }
    }

    private fun findSlidingPanel(): SlidingUpPanelLayout? {
        return supportFragmentManager.fragments
                .find { it is MainFragment }
                ?.view
                ?.findViewById(R.id.slidingPanel)
    }

}
